### PR TITLE
Pause on objective completion

### LINKED ISSUE #2073: Pause the game on objective completion
**Is your feature request related to a problem? Please describe.**

When an objective is completed, the game can keep running until the modal is shown.

In an extreme case, `--scenario Tutorials/move --autoplay` finishes the tutorial before the next modal pops up.

**Describe the solution you'd like**

Pause the game after objective completion and do not run further ticks.

### PR DESCRIPTION
**Pull request by xsebek**
* add a new option `--hide-goal  Do not show goal modal window that pauses the game.`
* pause the game once an objective is finished - can be disabled by `--hide-goal`
* show the goal modal window (and autopause) even when running with `--autoplay`
  * previous behavior can be achieved with `--autoplay --hide-goal`
  * removes the debug option `autoplay_goals`
* closes #2073

### CODE REVIEW

**Comment by xsebek** [comment]
For reference, this is how I debugged this with an output to file that showed the state of the objective display queue:
<details>
  <summary>Show patch and debug output</summary>

```diff
--- b/src/swarm-engine/Swarm/Game/Step.hs
+++ a/src/swarm-engine/Swarm/Game/Step.hs
@@ -341,9 +341,22 @@ hypotheticalWinCheck em g ws oc = do
     _ -> return ()
 
   queue <- messageInfo . announcementQueue Swarm.Util.<%= (>< Seq.fromList (map ObjectiveCompleted $ completionAnnouncementQueue finalAccumulator))
-  shouldPause <- use $ temporal . pauseOnCompletion
 
-  when (newWinState /= Ongoing || (notNull queue && shouldPause == PauseOnAnyObjective)) $
+  shouldPause <- use $ temporal . pauseOnCompletion
+  let willPause = newWinState /= Ongoing || (notNull queue && shouldPause == PauseOnAnyObjective)
+  -- TODO: remove this debug ouput
+  sendIO $
+    appendFile "log_win.txt" $
+      intercalate
+        " \t"
+        [ show $ getTickNumber ts
+        , if newWinState == Ongoing then "ongoing" else "won"
+        , if notNull queue then "queued" else "empty"
+        , show shouldPause
+        , if willPause then "AutoPause" else "Running"
+        ] <> "\n"
+
+  when willPause $
     temporal . runStatus .= AutoPause
 
   mapM_ handleException $ exceptions finalAccumulator
```
```
1 	ongoing 	empty 	PauseOnAnyObjective 	Running
2 	ongoing 	queued 	PauseOnAnyObjective 	AutoPause
3 	ongoing 	empty 	PauseOnAnyObjective 	Running
4 	ongoing 	empty 	PauseOnAnyObjective 	Running
5 	ongoing 	empty 	PauseOnAnyObjective 	Running
6 	ongoing 	empty 	PauseOnAnyObjective 	Running
7 	ongoing 	empty 	PauseOnAnyObjective 	Running
8 	ongoing 	queued 	PauseOnAnyObjective 	AutoPause
9 	ongoing 	empty 	PauseOnAnyObjective 	Running
10 	ongoing 	empty 	PauseOnAnyObjective 	Running
11 	ongoing 	empty 	PauseOnAnyObjective 	Running
12 	ongoing 	empty 	PauseOnAnyObjective 	Running
13 	ongoing 	queued 	PauseOnAnyObjective 	AutoPause
14 	ongoing 	empty 	PauseOnAnyObjective 	Running
15 	ongoing 	empty 	PauseOnAnyObjective 	Running
16 	ongoing 	empty 	PauseOnAnyObjective 	Running
17 	ongoing 	empty 	PauseOnAnyObjective 	Running
18 	ongoing 	empty 	PauseOnAnyObjective 	Running
19 	ongoing 	empty 	PauseOnAnyObjective 	Running
20 	ongoing 	empty 	PauseOnAnyObjective 	Running
21 	ongoing 	empty 	PauseOnAnyObjective 	Running
22 	ongoing 	empty 	PauseOnAnyObjective 	Running
23 	ongoing 	empty 	PauseOnAnyObjective 	Running
24 	ongoing 	empty 	PauseOnAnyObjective 	Running
25 	ongoing 	empty 	PauseOnAnyObjective 	Running
26 	ongoing 	empty 	PauseOnAnyObjective 	Running
27 	ongoing 	empty 	PauseOnAnyObjective 	Running
28 	ongoing 	empty 	PauseOnAnyObjective 	Running
29 	ongoing 	empty 	PauseOnAnyObjective 	Running
30 	ongoing 	empty 	PauseOnAnyObjective 	Running
31 	ongoing 	empty 	PauseOnAnyObjective 	Running
32 	ongoing 	empty 	PauseOnAnyObjective 	Running
33 	ongoing 	empty 	PauseOnAnyObjective 	Running
34 	ongoing 	empty 	PauseOnAnyObjective 	Running
35 	ongoing 	empty 	PauseOnAnyObjective 	Running
36 	ongoing 	empty 	PauseOnAnyObjective 	Running
37 	won 	queued 	PauseOnAnyObjective 	AutoPause
```
</details>

File: src/swarm-engine/Swarm/Game/Step.hs [hunk 1]
```diff
@@ -338,7 +339,11 @@ hypotheticalWinCheck em g ws oc = do
     Unwinnable _ -> grantAchievement LoseScenario
     _ -> return ()
 
-  messageInfo . announcementQueue %= (>< Seq.fromList (map ObjectiveCompleted $ completionAnnouncementQueue finalAccumulator))
+  queue <- messageInfo . announcementQueue Swarm.Util.<%= (>< Seq.fromList (map ObjectiveCompleted $ completionAnnouncementQueue finalAccumulator))
+  shouldPause <- use $ temporal . pauseOnCompletion
+
+  when (newWinState /= Ongoing || (notNull queue && shouldPause == PauseOnAnyObjective)) $
```
Comment (byorgey) [inline src/swarm-engine/Swarm/Game/Step.hs:345]: I'm trying to make sure I understand the logic here.  What is the purpose of the `notNull queue` check?  How does the `newWinState` check and the other check fit together?  (And can you turn your answer into some comments in the code here?)

File: src/swarm-tui/Swarm/TUI/Model/StateUpdate.hs [hunk 2]
```diff
@@ -127,8 +127,16 @@ initPersistentState ::
...
+    rs <-
+      initRuntimeState
+        RuntimeOptions
+          { startPaused = pausedAtStart
+          , pauseOnObjectiveCompletion = autoShowObjectives
+          , loadTestScenarios = Set.member LoadTestingScenarios debugOptions
+          }
+    let showMainMenu = not (skipMenu opts)
+    ui <- initUIState UIInitOptions {..}
+    -- \$ speed (not (skipMenu opts)) debugOptions
```
Comment (byorgey) [inline src/swarm-tui/Swarm/TUI/Model/StateUpdate.hs:139]: Seems like stray dead code --- should this be removed?

**Review by byorgey — APPROVED** [review]

File: src/swarm-engine/Swarm/Game/State/Runtime.hs [hunk 3]
```diff
@@ -79,32 +80,42 @@ initGameStateConfig ::
...
+  { startPaused :: Bool
+  , pauseOnObjectiveCompletion :: Bool
   , loadTestScenarios :: Bool
   }
   deriving (Eq, Show)
 
+instance Semigroup RuntimeOptions where
+  a <> b =
+    RuntimeOptions
+      (a.startPaused || b.startPaused)
```
[the remark this answers is not in the document]
Comment (xsebek) [reply src/swarm-engine/Swarm/Game/State/Runtime.hs]: I like it, too; it's nice that it is not ambiguous when the same name is in the scope.

The instance is not really needed though, so I removed it.

File: src/swarm-engine/Swarm/Game/State/Substate.hs [hunk 4]
```diff
@@ -274,11 +276,15 @@ data SingleStep
...
 
+data PauseOnObjective = PauseOnWin | PauseOnAnyObjective
+  deriving (Eq, Ord, Show, Enum, Bounded)
+
 data TemporalState = TemporalState
   { _gameStep :: Step
   , _runStatus :: RunStatus
   , _ticks :: TickNumber
   , _robotStepsPerTick :: Int
+  , _pauseOnCompletion :: PauseOnObjective
```
[the remark this answers is not in the document]
Comment (xsebek) [reply src/swarm-engine/Swarm/Game/State/Substate.hs]: I renamed one, but not the other. Thanks for spotting it. 👍

File: src/swarm-engine/Swarm/Game/Step.hs (hunk 1 above)
Comment (xsebek) [reply src/swarm-engine/Swarm/Game/Step.hs]: I named the parts using `let`. Is it clearer now? 🙂

File: src/swarm-tui/Swarm/TUI/Model/StateUpdate.hs (hunk 2 above)
Comment (xsebek) [reply src/swarm-tui/Swarm/TUI/Model/StateUpdate.hs]: Yes, these were the original parameters passed to `initUIState`. Now, only the `showMainMenu` needs to be declared, and the others flow from one record to another.

File: src/swarm-engine/Swarm/Game/Step.hs (hunk 1 above)
Comment (byorgey) [reply src/swarm-engine/Swarm/Game/Step.hs]: Yes, much better, thanks!

**Pull request merged**